### PR TITLE
[with_data_parallel][part5] remove with_data_parallel in unit test

### PR DESCRIPTION
**Pull request by kangguangli**
### PR types

Others
### PR changes

Others
### Describe

由于旧执行器预计在2.5版本退场，相关接口需要提前下线。本PR旨在去除`test_dist_base`中的`with_data_parallel`。

部分相关单测利用了`with_data_parallel`来插入通信算子，这里会改为通过`raw_program_optimizer`插入通信算子。
然而由于`DistributeTranspiler`会在`startup_program`中插入`gen_nccl_id`，这跟`raw_program_optimizer`初始化通信组的逻辑有冲突，会导致程序hang住。因此复制并修改了`raw_program_optimizer`的实现，去掉了插入通信组的部分。

As the `PE` is expected to be removed in version 2.5, relevant APIs need to be deleted in advance. This PR is intended to remove `with_data_parallel`  in `test_dist_base`。

Some unit tests use `with_data_parallel` to insert the communication op, which is changed to using `raw_program_optimizer`. However, because `DistributeTranspiler` will insert `gen_nccl_id` in `startup_program` , it is conflicted with the logic of `raw_program_optimizer` to init comm group, and cause the program to hang.  So copied and modified implementation of `raw_program_optimizer` to avoid this.

#### 其他修改 Other Changes
1. dist_mnist_dgc.py：`dgc_optimizer.py`初始化`dgc_op`时`rank`设为-1，由`sparse_all_reduce_op_handle`运行时设置正确的值。这里改为初始化时就设为`num_trainer`。以及去除了`num_trainer=1`下多卡case。
2. dist_allreduce_op.py：`dist_allreduce_op`改为通过`raw_program_optimizer`插入通信算子。
3. test_dist_base.py::run_trainer：
     - 调整了`BuildStrategy`初始化顺序，改到最前，以便传入给`dgc_optimizer.py`
     - 重定向了标准输出，确保标准输出只有最后的`out_losses`
     - 去除了`with_data_parallel`
 4. 其他

#### 其他 Others
Pcard-67004

### CODE REVIEW

File: python/paddle/fluid/compiler.py [hunk 1]
```diff
@@ -417,11 +417,11 @@ def _compile_data_parallel(self, places, use_device, scope=None):
...
-        if self._build_strategy.num_trainers > 1:
-            assert self._is_data_parallel, (
-                "If you use multi-trainer to train the model, you should use "
-                "the data parallel model, i.e. calling with_data_parallel function."
-            )
+        # if self._build_strategy.num_trainers > 1:
+        #     assert self._is_data_parallel, (
+        #         "If you use multi-trainer to train the model, you should use "
+        #         "the data parallel model, i.e. calling with_data_parallel function."
+        #     )
```
Comment (haohongxiang) [inline python/paddle/fluid/compiler.py]: Please delete these commented codes.

File: python/paddle/fluid/tests/unittests/dist_allreduce_op.py [hunk 2]
```diff
@@ -109,13 +110,18 @@ def get_model(self, batch_size=2, single_device=False):
...
+            fleet.init(strategy=strategy)
+            optimizer = fleet.distributed_optimizer(opt)
+            optimizer.minimize(avg_cost)
+            # params_grads = opt.backward(avg_cost)
+            # data_parallel_param_grads = []
+            # for p, g in params_grads:
+            #     # NOTE: scale will be done on loss scale in multi_devices_graph_pass using nranks.
+            #     grad_reduce = fluid.layers.collective._allreduce(g)
+            #     data_parallel_param_grads.append([p, grad_reduce])
+            # opt.apply_gradients(data_parallel_param_grads)
```
Comment (haohongxiang) [inline python/paddle/fluid/tests/unittests/dist_allreduce_op.py]: Same as above, please delete these commented codes.

File: python/paddle/fluid/compiler.py (hunk 1 above)
Comment (kangguangli) [reply python/paddle/fluid/compiler.py]: Done. Thanks.

File: python/paddle/fluid/tests/unittests/dist_allreduce_op.py (hunk 2 above)
Comment (kangguangli) [reply python/paddle/fluid/tests/unittests/dist_allreduce_op.py]: Done. Thanks.

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by wuhuachaocoding — APPROVED** [review]

**Review by ForFishes — APPROVED** [review]
LGTM

**Review by haohongxiang — APPROVED** [review]
LGTM

**Pull request merged**